### PR TITLE
Add Bayesian process analysis

### PR DESCRIPTION
**Pull request by maltelueken**
Adds the Bayesian process analysis with some integration tests and help files.

Reuses many functions from the classic process analysis.

Refactors some R functions so that they can be also used in the Bayesian analysis.

Extracts many shared QML elements into separate QML files stored in the `common/` folder.

Reuses MCMC plotting functions from the `jaspJags` module.

Adapts two functions from the `blavaan` package to slightly modify their behavior (necessary for LOO computation).

Installs `blavaan` from nlesc-jasp-projects/blavaan because the latest version has a bug which breaks the analysis when there is a moderation effect.

### CODE REVIEW

**Review by ThijsVroegh — CHANGES_REQUESTED** [review]
There are several issues similar to the ones raised in the PR on the covariances (because its integrated in this PR). I nonetheless mention them here again:

* Conceptual Model dissapears when adding an extra line with input for the model

* Shouldn't model output be made visible per default? Now it is collapsed, and need to be opened first to make it visible

* There are no parameter estimates in the statistical plot

* Residual variances only show the arrow but no parameter estimate for the variance

* There are sometimes 'left-over', empty tables while a model is being changed. 

* Is starting with an empty table (on model summary) necessary? Perhaps nicer when it emerges after the model has been defined. In terms of ordering; would it be possible to have the model summary directly under the conceptual path plots?

File: R/bayesianProcess.R [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+.procBayesResultsFitModel <- function(container, dataset, options, modelOptions) {
+  # Should model be fitted?
+  doFit <- .procCheckFitModel(container[["graph"]]$object)
+
+  if (!doFit) {
+    dataset <- NULL
+  }
+
+  # Necessary for JASP to find function blavaan
+  blavaan <- blavaan::blavaan
```
Comment (ThijsVroegh) [inline R/bayesianProcess.R:118]: You explained the reason for doing this. Is this a bug on the jasp side or on the current version of blavaan?

File: R/classicProcess.R [hunk 2]
```diff
@@ -1575,7 +1578,7 @@ ClassicProcess <- function(jaspResults, dataset = NULL, options) {
     if (is.null(container[[modelNames[i]]])) {
       modelContainer <- createJaspContainer(title = modelNames[i], , initCollapsed = TRUE)
       modelContainer$dependOn(
-        options = c("parameterLabels", "ciLevel"),
```
Comment (ThijsVroegh) [inline R/classicProcess.R:1544]: Why is this deleted?

File: R/classicProcess.R [hunk 3]
```diff
@@ -1742,6 +1745,51 @@ ClassicProcess <- function(jaspResults, dataset = NULL, options) {
   return(modProbes)
 }
 
+.procPathMediationEffectsTableHelper <- function(tbl, medEffects) {
```
Comment (ThijsVroegh) [inline R/classicProcess.R:1748]: this function takes mediation effects data and organizes it into a structured table format, preparing it for further analysis or visualization. How was this done before the addition of this function?

File: R/bayesianProcess.R (hunk 1 above)
Comment (maltelueken) [reply R/bayesianProcess.R:124]: It seems to be a bug in `blavaan` specific to JASP. When you run `blavaan` in R, it does not occurr.
The reason is that the `blavaan::bsem` function calls `blavaan` (not `blavaan::blavaan`) under the hood but JASP cannot find the function because it has not loaded the package (as you would do in a normal R session).

File: R/classicProcess.R [hunk 4]
```diff
@@ -76,8 +76,11 @@ ClassicProcess <- function(jaspResults, dataset = NULL, options) {
 # Helper function for generic dependencies of all models
 .procGetDependencies <- function() {
   return(c(
-    "dependent", "covariates", "factors", "naAction", "emulation", "estimator", "meanCenteredModeration",
-    "standardizedModelEstimates", "errorCalculationMethod", "bootstrapCiType"
+    "dependent", "covariates", "factors", "naAction", "emulation", "estimator",
+    "meanCenteredModeration", "standardizedModelEstimates", "errorCalculationMethod", "bootstrapCiType",
+    "mcmcBurnin", "mcmcSamples", "mcmcChains", "seed", "setSeed", "nuPriorMu",
+    "nuPriorSigma", "betaPriorMu", "betaPriorSigma", "psiPriorAlpha",
+    "psiPriorBeta", "rhoPriorAlpha", "rhoPriorBeta"
```
[the remark this answers is not in the document]
Comment (maltelueken) [reply R/classicProcess.R:84]: They run independently, but `BayesianProcess` reuses many functions from `ClassicProcess` (including the GUI dependencies), so it was much easier to add the Bayesian-specific GUI dependencies there. In theory, you are right, it would be better to have them in a separate function, but then we would need to adapt many of the existing functions (which I didn't do in the interest of time).

File: R/classicProcess.R (hunk 3 above)
Comment (maltelueken) [reply R/classicProcess.R:1798]: Before this was done in the `.procPathMediationEffectsTable` function itself, but now the Bayesian analysis is also using the same routine, so I put it into a separate function.

File: inst/help/BayesianProcess.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+- Dependent: The final dependent variable of interest.
+- Continuous Predictors: Continuous or ordinal predictor variables.
+- Categorical Predictors: Categorical independent predictor variables.
+
+#### Models
+
+This section allows users to specify multiple process models through one of two interfaces. New models can be added by clicking the green `+` button. For each model, the following options are available:
+
+- Input type:
+	- Variables: Models can be specified by adding and specifying paths. Paths indicate causality. For each path, there are four options
```
Comment (ThijsVroegh) [inline inst/help/BayesianProcess.md:48]: change 'variables' into 'paths' so that it matches the recent updated naming

File: inst/help/BayesianProcess.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		- From: The variable from which the path starts.
+		- To: The variable at which the path ends.
+		- Process Type: The type process influencing the path.
+			- Mediator: The path is mediated by a third variable. Adds an indirect path through the `Process Variable`.
+			- Moderator: The path is moderated by a third variable. Adds a direct path from the `Process Variable` and the interaction between the `Process Variable` and `From` to the `To` variable.
+			- Confounder: The path is confounded by a third variable. Adds a direct path from the `Process Variable` both to `From` and `To`.
+			- Direct: Adds a direct path from `From` to `To`.
+	- Model number: Models can be specified by choosing a `Hayes model number` and adding the required variables to assignment boxes.
```
Comment (ThijsVroegh) [inline inst/help/BayesianProcess.md:56]: likewise, change 'model number' into 'Hayes configuration'. This is perhaps also a good place to make a clear reference where the user can get more knowledge on the different Hayes models (i.e., the book and its appendix)

File: inst/help/BayesianProcess.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		- Moderator Z: Second moderator variable.
+
+- Residual Covariances: Adds covariances between the residuals of variables indicating that they have a common cause not specified in the model. Also adds variances for independent (exogenous) variables. This adds degrees of freedom corresponding to the number of covariances and variances added.
+- Parameter Estimates: Determines which parameter estimates are shown in the output. Requesting intercepts adds the mean structure to the model, which adds parameters and will therefore affect some fit indices (such as the information criteria AIC and BIC).
+- Path plots: Displays path plots in the output.
+	- Conceptual: Shows a conceptual path plot. These plots contain no parameter labels or estimates and simplify moderation effects in the model structure by omitting direct effects and interactions. Instead moderators point to the path they are moderating. These plots can be shown even when the model is not complete yet (when `Input type` is `Model number`) and are easier to understand than the statistical path plots.
+	- Statistical: Shows a statistical path plot including parameter labels or estimates for all paths.
+
+#### Options
```
Comment (ThijsVroegh) [inline inst/help/BayesianProcess.md:68]: perhaps restructure this, following the order in the tab

File: inst/help/BayesianProcess.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	- Density: Show a density plot of the posterior samples.
+	- Histogram: Show a histogram of the posterior samples.
+	- Trace: Show a trace plot of the posterior samples.
+	- Autocorrelation: Plot the autocorrelation of the posterior samples.
+		- No. lags: the maximum number of lags to show in the autocorrelation plot.
+		- Type: whether to display the autocorrelation as a bar at each lag, or as a line that connects subsequent lags.
+	- Bivariate scatter: Show a bivariate scatter plot of all pairs of variables. Only shows output when more than 1 parameter is sampled.
+		- Diagonal plot type: Show a density plot or a histogram on the diagonal entries of the scatter plot.
+		- Off-diagonal plot type: Show a hexagonal bivariate density plot, or a contour plot on the off-diagonal entries of the scatter plot.
+- Path plots: Options for path plots.
```
Comment (ThijsVroegh) [inline inst/help/BayesianProcess.md:93]: move this part above the MCMC explanation, following the structure in the plot tab

File: inst/help/BayesianProcess.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+- All univariate distributions of dependent variables should be normal.
+- All bivariate associations between dependent variables are linear.
+- The distribution of residuals does not depend on the independent variables, i.e., they are homoscedastic.
+
+### Input
+---
+#### Assignment Box
+
+- Dependent: The final dependent variable of interest.
```
[the remark this answers is not in the document]
Comment (maltelueken) [reply inst/help/BayesianProcess.md]: Good idea, I added that!

**Comment by maltelueken** [comment]
I updated the help files according to your suggestions! I will look into the Bayesian issue later.

**Comment by maltelueken** [comment]
The tests run fine locally, I need to look more into why they are failing online, but I will do that in a separate PR. Issue #89 is more elaborate, so I will also tackle it in a separate PR.

**Review by ThijsVroegh — APPROVED** [review]
All comments have been addressed. Help file is now structured following the tabs in the module, making it an easy read and of high value for newcomers working with the Bayesian module. The earlier made issue is more complex and will be addressed in a separate PR. An additional (but probably minor) issue was filed on the usage of the color pallette, which only seems to work for the statistical path plot.

**Comment by JorisGoosen** [comment]
Aaah, I just removed `nominalText` from everything...

And this PR adds it back...

Could it be removed again?
It doesnt exist anymore in JASP so there is little point in having it there

**Comment by JorisGoosen** [comment]
Dank je!
